keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt

```diff
@@ -0,0 +1,386 @@
+ * store artifacts and resources.
+ *
+ * This also gives us an easy place to emit telemetry and events around the usage of methods.
+ *
+ * TODO eb: refactor repository interaction so transactionality is easier.
+ */
+@Component
+class CombinedRepository(
+  val deliveryConfigRepository: DeliveryConfigRepository,
+  val artifactRepository: ArtifactRepository,
+  val resourceRepository: ResourceRepository,
+  val clock: Clock,
+  val publisher: ApplicationEventPublisher
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  fun upsertDeliveryConfig(submittedDeliveryConfig: SubmittedDeliveryConfig): DeliveryConfig {
+    val new = DeliveryConfig(
+      name = submittedDeliveryConfig.name,
+      application = submittedDeliveryConfig.application,
+      serviceAccount = submittedDeliveryConfig.serviceAccount,
+      artifacts = submittedDeliveryConfig.artifacts.transform(submittedDeliveryConfig.name),
+      environments = submittedDeliveryConfig.environments.mapTo(mutableSetOf()) { env ->
+        Environment(
+          name = env.name,
+          resources = env.resources.mapTo(mutableSetOf()) { resource ->
+            resource
+              .copy(metadata = mapOf("serviceAccount" to submittedDeliveryConfig.serviceAccount) + resource.metadata)
+              .normalize()
+          },
+          constraints = env.constraints,
+          notifications = env.notifications
+        )
+      }
+    )
+
+    validate(new)
+    return upsertDeliveryConfig(new)
+  }
+
+  /**
+   * Validates that resources have unique ids, throws an exception if invalid
+   */
+  private fun validate(config: DeliveryConfig) {
+    val resources = config.environments.map { it.resources }.flatten().map { it.id }
+    val distinct = resources.distinct()
+
+    if (resources.size != distinct.size) {
+      val duplicates = resources.groupingBy { it }.eachCount().filter { it.value > 1 }.keys.toList()
+      val envToResources: Map<String, MutableList<String>> = config.environments
+        .map { env -> env.name to env.resources.map { it.id }.toMutableList() }.toMap()
+      val envsAndDuplicateResources = envToResources
+        .filterValues { rs: MutableList<String> ->
+          // remove all the resources we don't care about from this mapping
+          rs.removeIf { it !in duplicates }
+          // if there are resources left that we care about, leave it in the map
+          rs.isNotEmpty()
+        }
+      log.error("Validation failed for ${config.name}, duplicates found: $envsAndDuplicateResources")
+      throw DuplicateResourceIdException(duplicates, envsAndDuplicateResources)
+    }
+  }
+
+  private fun Set<DeliveryArtifact>.transform(deliveryConfigName: String) =
+    map { artifact ->
+      when (artifact) {
+        is DockerArtifact -> artifact.copy(deliveryConfigName = deliveryConfigName)
+        is DebianArtifact -> artifact.copy(deliveryConfigName = deliveryConfigName)
+      }
+    }.toSet()
+
```

Comment:
There's a problem in this function `upsertDeliveryConfig(deliveryConfig: DeliveryConfig)` - we're not actually rolling back when there's a problem (or, the test is doing the wrong thing now that we have different imports). Would love some eyes on this.